src/tree/hist/histogram.h

```diff
@@ -180,47 +214,100 @@ class HistogramBuilder {
 
     common::BlockedSpace2d space(
         nodes_to_build.size(), [&](std::size_t) { return n_total_bins; }, 1024);
-    common::ParallelFor2d(space, this->n_threads_, [&](size_t node, common::Range1d r) {
+    common::ParallelFor2d(space, this->n_threads_, [&](std::size_t node, common::Range1d r) {
       // Merging histograms from each thread.
       this->buffer_.ReduceHist(node, r.begin(), r.end());
     });
     if (is_distributed_ && !is_col_split_) {
-      // The cache is contiguous, we can perform allreduce for all nodes in one go.
+      // Horizontal
       CHECK(!nodes_to_build.empty());
-      auto first_nidx = nodes_to_build.front();
-      std::size_t n = n_total_bins * nodes_to_build.size() * 2;
-      collective::Allreduce<collective::Operation::kSum>(
+      if (!is_secure_) {
+        // Non-secure mode, we directly perform allreduce
+        // The cache is contiguous, we can perform allreduce for all nodes in one go.
+        auto first_nidx = nodes_to_build.front();
+        std::size_t n = n_total_bins * nodes_to_build.size() * 2;
+        collective::Allreduce<collective::Operation::kSum>(
           reinterpret_cast<double *>(this->hist_[first_nidx].data()), n);
+      } else {
+        // Secure mode, we need to call interface to perform encryption and decryption
+        // note that the actual aggregation will be performed at server side
+        auto first_nidx = nodes_to_build.front();
+        std::size_t n = n_total_bins * nodes_to_build.size() * 2;
+        auto hist_to_aggr = std::vector<double>();
+        for (std::size_t hist_idx = 0; hist_idx < n; hist_idx++) {
+          double hist_item = reinterpret_cast<double *>(this->hist_[first_nidx].data())[hist_idx];
+          hist_to_aggr.push_back(hist_item);
+        }
+        std::size_t size;
+        auto hist_buf = processor_instance->ProcessHistograms(&size, hist_to_aggr);
+        std::size_t buffer_size{};
+        std::int8_t *buffer;
+        buffer_size = size;
+        buffer = reinterpret_cast<std::int8_t *>(hist_buf);
+        auto hist_vec = std::vector<std::int8_t>(buffer, buffer + buffer_size);
+        auto hist_entries = collective::AllgatherV(hist_vec);
+        std::vector<double> hist_aggr =
+                processor_instance->HandleHistograms(hist_entries.data(), hist_entries.size());
+        // Assign the aggregated histogram back to the local histogram
+        for (std::size_t hist_idx = 0; hist_idx < n; hist_idx++) {
+          reinterpret_cast<double *>(this->hist_[first_nidx].data())[hist_idx] =
+                  hist_aggr[hist_idx];
+        }
+      }
     }
 
     if (is_distributed_ && is_col_split_ && is_secure_) {
-      // Under secure vertical mode, we perform allgather for all nodes
+      // Under secure vertical mode, we perform allgather to get the global histogram.
```

Comment:
Let's focus on horizontal as described by the PR

Reply:
same as other vertical comments, after fixing and merging https://github.com/dmlc/xgboost/pull/10124, this PR should look much cleaner